exporter/sumologicexporter/sender.go

```diff
@@ -55,23 +56,38 @@ func newAppendResponse() appendResponse {
 	}
 }
 
-func newSender(ctx context.Context, cfg *Config, cl *http.Client, f filter) *sender {
+func newSender(ctx context.Context, cfg *Config, cl *http.Client, f filter, c compressor) *sender {
 	return &sender{
-		config: cfg,
-		client: cl,
-		filter: f,
-		ctx:    ctx,
+		config:     cfg,
+		client:     cl,
+		filter:     f,
+		ctx:        ctx,
+		compressor: c,
 	}
 }
 
 // send sends data to sumologic
 func (s *sender) send(pipeline PipelineType, body io.Reader, flds fields) error {
-	// Add headers
-	req, err := http.NewRequestWithContext(s.ctx, http.MethodPost, s.config.HTTPClientSettings.Endpoint, body)
+	data, err := s.compressor.compress(body)
```

Comment:
This is a bit inefficient because you writhe the body once then read again to compress. I think you can do compression on the fly, but I may be wrong

Reply:
I'm not a go optimization expert, but only pattern I found in standard library uses a goroutine and pipe

https://gist.github.com/tomcatzh/cf8040820962e0f8c04700eb3b2f26be

Maybe the copy is fine instead of syncing between routines.

Reply:
@pmalek-sumo made some benchmarking of the both solutions, and copy was better for this case

Reply:
Allow me to add what I've been able to test as @sumo-drosiek mentioned:

I've added the following alternative implementation of `compress()`

```go
func (c *compressor2) compress(data io.Reader) io.Reader {
	r, w := io.Pipe()

	go func(data io.Reader) {
		c.writer.Reset(w)
		defer func() {
			if err := w.Close(); err != nil {
				log.Errorf("failed closing write end of pipe: %v", err)
			}
		}()

		_, err := io.Copy(c.writer, data)
		if err != nil {
			log.Errorf("failed copying to pipe: %v", err)
			return
		}

		if err = c.writer.Close(); err != nil {
			log.Errorf("failed closing gzip writer end of pipe: %v", err)
			return
		}
	}(data)

	return r
}
```

and I've run the following benchmarks with this and the original implementation (from this PR):

```go
package main

import (
	"compress/gzip"
	"flag"
	"fmt"
	"io"
	"io/ioutil"
	"os"
	"strings"
	"testing"

	log "github.com/sirupsen/logrus"
)

const text = "aaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaa"

var print = flag.Bool("print", false, "print buffers")

func BenchmarkGZIP_PR(b *testing.B) {
	b.StopTimer()
	comp, err := newCompressor(GZIPCompression)
	if err != nil {
		panic(err)
	}
	str := strings.NewReader(text)
	b.StartTimer()

	for i := 0; i < b.N; i++ {
		b.StopTimer()
		if _, err := str.Seek(0, io.SeekStart); err != nil {
			panic(err)
		}
		b.StartTimer()

		c, err := comp.compress(str)
		if err != nil {
			log.Errorf("failed on compress(): %v", err)
			continue
		}

		if *print {
			printStdout(c)
		} else {
			_, err = io.Copy(ioutil.Discard, c)
			if err != nil {
				log.Errorf("failed on io.Copy(): %v", err)
			}
		}
	}
}

func BenchmarkGZIP_new(b *testing.B) {
	b.StopTimer()
	comp, err := newCompressor2(GZIPCompression)
	if err != nil {
		panic(err)
	}
	str := strings.NewReader(text)
	b.StartTimer()

	for i := 0; i < b.N; i++ {
		b.StopTimer()
		if _, err := str.Seek(0, io.SeekStart); err != nil {
			panic(err)
		}
		b.StartTimer()

		c := comp.compress(str)

		if *print {
			printStdout(c)
		} else {
			_, err = io.Copy(ioutil.Discard, c)
			if err != nil {
				log.Errorf("failed on io.Copy(): %v", err)
			}
		}
	}
}

func printStdout(r io.Reader) {
	cc, err := gzip.NewReader(r)
	if err != nil {
		log.Errorf("failed on gzip.NewReader: %v", err)
	}
	if _, err = io.Copy(os.Stdout, cc); err != nil {
		log.Errorf("failed on io.Copy(): %v", err)
		return
	}
	fmt.Println()
	if err = cc.Close(); err != nil {
		log.Errorf("failed on cc.Close(): %v", err)
	}
}
```

And I'm getting pretty similar results between those 2:

On MBP

```
go test -v . -count=1 -bench=BenchmarkGZIP -benchmem -benchtime=3s
goos: darwin
goarch: amd64
pkg: gzip
BenchmarkGZIP_PR
BenchmarkGZIP_PR-16                31280            111129 ns/op         1378663 B/op         16 allocs/op
BenchmarkGZIP_new
BenchmarkGZIP_new-16               28630            123268 ns/op         1394925 B/op         21 allocs/op
PASS
ok      gzip    15.457s
```

Using docker and `golang:1.15.6` image:

```
docker run --rm -it -v $(pwd):/app -w /app golang:1.15.6 bash
Status: Downloaded newer image for golang:1.15.6
root@b0aa3e99ade3:/app# go test -v . -count=1 -bench=BenchmarkGZIP -benchmem -benchtime=3s
go: downloading github.com/klauspost/pgzip v1.2.5
go: downloading github.com/sirupsen/logrus v1.7.0
go: downloading golang.org/x/sys v0.0.0-20191026070338-33540a1f6037
go: downloading github.com/klauspost/compress v1.11.3
goos: linux
goarch: amd64
pkg: gzip
BenchmarkGZIP_PR
BenchmarkGZIP_PR-11                16035            211124 ns/op         1459514 B/op         17 allocs/op
BenchmarkGZIP_new
BenchmarkGZIP_new-11               15700            232808 ns/op         1463768 B/op         21 allocs/op
PASS
ok      gzip    28.703s
```